homeassistant/components/sma/sensor.py

```diff
@@ -166,6 +168,8 @@ def __init__(
         self._config_entry_unique_id = config_entry_unique_id
         self._device_info = device_info
 
+        self._attr_entity_category = self._sensor.entity_category
```

Comment:
This is not information that should come out of the library, but instead should be defined here in the integration level; as this is not a device property.

Reply:
All the sensors/entities that are part of the device are generated by the integration. In the definitions in the integration I added the option to specify the `entity_category`. This value can thus be different for each sensor and is not a device property.

Reply:
The problem is that the definition is outside of Home Assistant:

<https://github.com/kellerza/pysma/compare/0.6.7...0.6.8#diff-36324cee262f97208c412d9d37fb31edf5509ad18cf9544ce1848df195449741R30>

It should be in the integration, not outside of it.

Reply:
Maybe I don't understand the use of `entity_category` than. Can you elaborate on this requirement?

The way the `pysma` library was build the sensors are defined by the library and a `SensorEntity` is created within the integration for each `Sensor` the library provides. I don't see a way, with the current setup, to define the entity category at the integration level.

Reply:
Move the definition to Home Assistant. We have EntityDescription classes to help out with that.

Reply:
If you can point me to documentation for the use of `EntityDescription` I will have a look at it.

Reply:
Searching... We actually don't have docs on that! Darn, that is bad :(
If you search for EntityDescription in this repository, you'll find tons of examples though.

Reply:
I looked up some examples, but this just moves some parameters of the entity to the `EntityDescription` class. I like the cleanliness of it, but I would still require the `pysma` library to provide the correct sensors for the device type it is connecting to.

The other option would be to move all sensor definitions and the logic that determines the type of SMA device and sensors associated with it from the library to the integration. This however requires a major change to both the library and the integration.

Reply:
> This however requires a major change to both the library and the integration.

Yes.

Reply:
Unfortunately I don't have the time to do a complete rewrite of the integration and library. Also, I don't agree with moving the definitions and logic to the integration as that makes the library less useful for use outside of home assistant.

If you want to determine the entity category in home assistant we could change the property of the sensor definition in the library to `is_diagnostic` and set the category from within home assistant as such:
```python
        if self._sensor.is_diagnostic:
            self._attr_entity_category = ENTITY_CATEGORY_DIAGNOSTIC
```
This is however less scalable if more categories are added in the future. A property would be needed for each category.

Reply:
> Also, I don't agree with moving the definitions and logic to the integration as that makes the library less useful for use outside of home assistant.

You don't have to move them out of the library, you can also map them explicitly from within Home Assistant to the library.

In the end, it is important we have the source of the entities in Home Assistant.

Reply:
Trying to understand how you see this. We keep the Sensor definitions in the library and move the HA specific parameters to a dictionary of `EntityDescription` objects? Then as soon as a sensor is created in home assistant we look up the matching `EntityDescription` en pass that to the entity?

Reply:
@frenck, I have implemented `SensorEntityDescription`. Let me know if this is what you had in mind. See 7321097dc4dd3ad7160b5198914e8b06b77f0851

Also did an improvement on the `DeviceInfo` (ece051324d5344173744d650245cb475df284a4a)

Reply:
Seeing such a big list, I think that you would be better off using less EntityDescription, and make them more generic:
- turn SENSOR_ENTITIES into a dict[key, SensorEntityDescription] so that you don't have to loop through the list to find the correct entity
- define the sensor name outside of the EntityDescription (using self._attr_name during initialisation)
- which allows you to make the key more generic (for example pv_power not pv_power_a/pv_power_b/pv_power_c)

Inside SMA, you could also define a "sensor_category" for each sensor that maps from SMA to the corresponding EntityCategory.

Reply:
Example (SMA):
```python
sensor.name = "power_l1"
sensor.category="power"
```

Would map to `SENSOR_ENTITTIES.get(sensor.category)`:
```python
SENSOR_ENTITTIES["power]= SensorEntityDescription(
        key="power",
        native_unit_of_measurement=POWER_WATT,
        state_class=STATE_CLASS_MEASUREMENT,
        device_class=DEVICE_CLASS_POWER,
        entity_registry_enabled_default=False,
    )
```

Reply:
The same category not always has the same settings. For example pv_power_a and pv_power_b are enable by default, but pv_power_c is not. I could remove the `entity_registry_enabled_default` from the `SensorEntityDescription` and use the old way using the `enabled` parameter from the `pysma.Sensor` class and set it in `SMAsensor.__init__()`:
```
self.entity_description.entity_registry_enabled_default = self._sensor.enabled
```